### DISCUSSION TITLE
Project Knowledge Base - Automatically Update Files When Git Branch Changes

### DISCUSSION THREAD

**Discussion by dallenpyrah** [open-webui/open-webui#7984] [Ideas]
# Feature Request

When working with OpenWebUI's knowledge base feature and Git repositories, the system doesn't automatically detect and update files when switching between different Git branches. This creates inconsistency between the actual codebase being referenced and the knowledge base content, leading to potential confusion and incorrect context during conversations.

Implement an automatic file update mechanism that:
- Monitors Git branch changes in connected repositories
- Automatically refreshes the knowledge base content when a branch switch is detected
- Updates file references and embeddings to match the current branch's content
- Provides a notification or status indicator when updates are in progress
- Maintains conversation history while updating the underlying knowledge base

This feature would be particularly valuable for development teams that:
- Work across multiple branches simultaneously
- Need to maintain accurate context when discussing different versions of code
- Use feature branches for development
- Require branch-specific documentation and context

The implementation should be efficient and avoid unnecessary reprocessing of unchanged files while ensuring the knowledge base stays synchronized with the current Git branch.

**Comment by flefevre** [comment]
Thanks for your proposal.
If I do understand we can put git url in knowledge base? Where exactly?
I was not aware of it, no mention in documentation.
Could you share some screenshots or documents?
Thanks

**Reply by dallenpyrah** [reply]
I don't think we can put GIT Url in knowledge base, but we can upload a local folder of our entire repository. 

I am thinking that if we can do that, what if we built a better solution for those who do want to upload that local folder of their repository?

**Reply by fnep** [reply]
Really being able to connect a git repository via its url (maybe even with credentials) would be awesome. I think as a workaround one could build a cronjob script to add a copy of the repo or relevant files to the knowledge base via api.
